Added new GRPC message types for worker indexing, along with relevant functions and tests

anandagopal6 opened the pull request:
### Protobuf needs to support sending and receiving worker metadata requests/responses

This PR adds the new worker indexing GRPC message types to the protobuf. It also adds some functions into GrpcWorkerChannel.cs to send/receive the metadata requests and responses, along with a test to check this sending and receiving functionality. 

Resolves #7513 and #7514 

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [x] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)

anandagopal6 committed:
Added new GRPC message types for worker indexing.

Added functions into GrpcWorkerChannel to handle sending requests and receiving responses form the worker.

Added test to check functionality of new worker channel functions.

anandagopal6 committed:
Update release_notes.md

brettsam commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -428,6 +438,69 @@ internal async Task SendInvocationRequest(ScriptInvocationContext context)
+                        .Take(1)
+                        .Subscribe((msg) => ProcessMetadata(msg.Message.WorkerMetadataResponse), HandleWorkerMetadataRequestError));
+
+            _workerChannelLogger.LogInformation("Sending WorkerMetadataRequest to {language} worker with worker ID {workerID}", _runtime, _workerId);
```
I think this should be `LogDebug` -- that way it will show up in our logs, but customers generally won't see it.

anandagopal6 replied:
Latest commit should address this

brettsam commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -428,6 +438,69 @@ internal async Task SendInvocationRequest(ScriptInvocationContext context)
+        // parse metadata response into FunctionMetadata objects
+        internal void ProcessMetadata(WorkerMetadataResponse workerMetadataResponse)
+        {
+            _workerChannelLogger.LogInformation("Received the worker response");
```
`LogDebug` here...

Also, make it a little more descriptive with `Received the worker metadata response.`.

anandagopal6 replied:
Latest commit should address this

brettsam commented on src/WebJobs.Script.Grpc/azure-functions-language-worker-protobuf/src/proto/FunctionRpc.proto:
```diff
@@ -75,6 +75,11 @@ message StreamingMessage {
     CloseSharedMemoryResourcesResponse close_shared_memory_resources_response = 28;
+
+    // Worker indexing message types
+    WorkerMetadataRequest worker_metadata_request = 29;
```
You all may have discussed the naming of these messages already... but should this be something like `WorkerFunctionMetadataRequest` (and `Response`) instead? Or even just `FunctionMetadataRequest`?

This makes it sound like it's metadata about the worker, rather than metadata about the functions.

anandagopal6 replied:
Latest commit should address this

brettsam commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -400,12 +403,19 @@ internal async Task SendInvocationRequest(ScriptInvocationContext context)
                 }
+                else if (_metadataRequestErrors.ContainsKey(context.FunctionMetadata.GetFunctionId()))
+                {
+                    _workerChannelLogger.LogDebug($"Function {context.FunctionMetadata.Name} failed to load");
```
Make this message slightly different from above -- like `... failed to load metadata from worker.`

anandagopal6 replied:
Latest commit should address this

brettsam commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -400,12 +403,19 @@ internal async Task SendInvocationRequest(ScriptInvocationContext context)
+                else if (_metadataRequestErrors.ContainsKey(context.FunctionMetadata.GetFunctionId()))
+                {
+                    _workerChannelLogger.LogDebug($"Function {context.FunctionMetadata.Name} failed to load");
+                    context.ResultSource.TrySetException(new Exception(_metadataRequestErrors[context.FunctionMetadata.GetFunctionId()].Message));
```
You shouldn't have to create a new `Exception` here -- can you just do `TrySetException(_metadataRequestErrors[context.FunctionMetadata.GetFunctionId()])`?

anandagopal6 replied:
The new `Exception` is needed unfortunately because `_metadataRequestErrors[context.FunctionMetadata.GetFunctionId()]` gives an `RpcException` while `context.ResultSource.TrySetException` requires something of type `Exception`. 

I originally had the line that you suggested but ran into issues because of the type difference.

brettsam replied:
Can you see how `_functionLoadErrors` handles this? I think that it may be changing that RpcException into an actual Exception before putting it in the collection. This should do the same thing, if possible.

anandagopal6 replied:
Good catch, turns out there is a method that checks the status and returns an `RpcException` (different `RpcException `than what I was defining).  This different `RpcException` can be saved as an `Exception`, even without casting, and that is how `_functionLoadErrors` handles it.

Making the change now!

anandagopal6 committed:
Switched some LogInformation calls to LogDebug calls and renamed new GRPC message types based on Brett's feedback

Also changed some debug message descriptiveness based on the feedback too.

anandagopal6 committed:
Changing exception types for metadata request exceptions that are thrown by the worker.

brettsam commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -428,6 +438,69 @@ internal async Task SendInvocationRequest(ScriptInvocationContext context)
+            {
+                if (metadata.Status != null && metadata.Status.IsFailure(out Exception metadataRequestEx))
+                {
+                    _metadataRequestErrors[metadata.Id] = metadataRequestEx;
```
Should we log an error here?

anandagopal6 replied:
Good call, just made the change!

brettsam reviewed (commented):
Just one more suggestion -- then looks good!

anandagopal6 committed:
Update GrpcWorkerChannel.cs

Added a debug log statement for when the worker indexing result for a function is a failure.

brettsam commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -428,6 +438,70 @@ internal async Task SendInvocationRequest(ScriptInvocationContext context)
+            {
+                if (metadata.Status != null && metadata.Status.IsFailure(out Exception metadataRequestEx))
+                {
+                    _workerChannelLogger.LogDebug($"Worker failed to index function {metadata.Id}");
```
Change this to `LogError()` -- then should be all good!

anandagopal6 replied:
Good call, just made the change!

brettsam reviewed (dismissed):
Just one minor change -- then we'll get this in!

anandagopal6 committed:
Update GrpcWorkerChannel.cs

Change log statement to a `LogError`

anandagopal6 committed:
Merge branch 'dev' into t-anjanan/grpc-additions
